zdm/tests/mc_statistics.py

```diff
@@ -0,0 +1,386 @@
+matplotlib.rc('font', **font)
+
+def main():
+
+	############## Initialise cosmology ##############
+	cos.init_dist_measures()
+	
+	# get the grid of p(DM|z). See function for default values.
+	# set new to False once this is already initialised
+	zDMgrid, zvals,dmvals=misc_functions.get_zdm_grid(
+		new=True,plot=False,method='analytic')
+	# NOTE: if this is new, we also need new surveys and grids!
+	
+	# constants of beam method
+	thresh=0
+	method=2
+	
+	
+	# sets which kind of source evolution function is being used
+	source_evolution=0 # SFR^n scaling
+	#source_evolution=1 # (1+z)^(2.7n) scaling
+	
+	
+	# sets the nature of scaling with the 'spectral index' alpha
+	alpha_method=0 # spectral index interpretation: includes k-correction. Slower to update
+	#alpha_method=1 # rate interpretation: extra factor of (1+z)^alpha in source evolution
+	
+	############## Initialise surveys ##############
+	
+	# constants of intrinsic width distribution
+	Wlogmean=1.70267
+	Wlogsigma=0.899148
+	DMhalo=50
+	
+	#These surveys combine time-normalised and time-unnormalised samples 
+	NewSurveys=False
+	#sprefix='Full' # more detailed estimates. Takes more space and time
+	sprefix='Std' # faster - fine for max likelihood calculations, not as pretty
+	
+	if sprefix=='Full':
+		Wbins=10
+		Wscale=2
+		Nbeams=[20,20,20]
+	elif sprefix=='Std':
+		Wbins=5
+		Wscale=3.5
+		Nbeams=[5,5,10]
+	
+	# location for survey data
+	sdir = os.path.join(resource_filename('zdm', 'data'), 'Surveys/')
+	if NewSurveys:
+		
+		print("Generating new surveys, set NewSurveys=False to save time later")
+		#load the lat50 survey data
+		lat50=survey.survey()
+		lat50.process_survey_file(sdir+'CRAFT_class_I_and_II.dat')
+		lat50.init_DMEG(DMhalo)
+		lat50.init_beam(nbins=Nbeams[0],method=2,plot=False,thresh=thresh) # tells the survey to use the beam file
+		pwidths,pprobs=survey.make_widths(lat50,Wlogmean,Wlogsigma,Wbins,scale=Wscale)
+		efficiencies=lat50.get_efficiency_from_wlist(dmvals,pwidths,pprobs)
+		
+		
+		# load ICS data
+		ICS=survey.survey()
+		ICS.process_survey_file(sdir+'CRAFT_ICS.dat')
+		ICS.init_DMEG(DMhalo)
+		ICS.init_beam(nbins=Nbeams[1],method=2,plot=False,thresh=thresh) # tells the survey to use the beam file
+		pwidths,pprobs=survey.make_widths(ICS,Wlogmean,Wlogsigma,Wbins,scale=Wscale)
+		efficiencies=ICS.get_efficiency_from_wlist(dmvals,pwidths,pprobs)
+		
+		# load Parkes data
+		pks=survey.survey()
+		pks.process_survey_file(sdir+'parkes_mb_class_I_and_II.dat')
+		pks.init_DMEG(DMhalo)
+		pks.init_beam(nbins=Nbeams[2],method=2,plot=False,thresh=thresh) # need more bins for Parkes!
+		pwidths,pprobs=survey.make_widths(pks,Wlogmean,Wlogsigma,Wbins,scale=Wscale)
+		efficiencies=pks.get_efficiency_from_wlist(dmvals,pwidths,pprobs)
+		
+		
+		names=['ASKAP/FE','ASKAP/ICS','Parkes/Mb']
+		
+		surveys=[lat50,ICS,pks]
+		if not os.path.isdir('Pickle'):
+			os.mkdir('Pickle')
+		with open('Pickle/'+sprefix+'surveys.pkl', 'wb') as output:
```

Comment:
Am curious -- is there a way around `pickle`?

These can be tough to handle across platforms..

Reply:
This is true. I use Windows, and usually, I need to change these parts of the code for them to work on my device.

Reply:
I don't know of a good way to save class objects with various internal arrays other than 'pickle', except manually saving e.g. each thing as various numpy arrays and loading individually.
Btw, does pickle just not work on Windows at all, or does what you read back in not equal what you put out? If you pickle on what platform and re-load on another this is expected to cause problems, but I'm surprised it's an issue within a platform. In any case, this pickle behaviour is fundamentally a speedup, so we might try making a universal flag titled "DoPickle" or some such that if turned to False just ignores the pickle load/save and generates everything fresh.

Reply:
Right, so path names and locations in Windows use the backslash \ instead of the normal slash /. This creates a problem wherein I usually need to go to the specific part in the code where the path is specified and make some other changes as well. I don't think this is specific to pickle, just a problem when using the same code across platforms. The universal flag idea does sound good.

Reply:
Ah OK! So it's essentially any part of the code that has a hard-coded '/' in it. I wonder how this is normally dealt with? A global variable called "appropriate_slash"

Reply:
the 'easy' way is to use `os.path.join`

e.g.  `os.path.join('Pickle', filename)`